Minibatch with multiple variables is broken

### Description

Found out in https://discourse.pymc.io/t/how-to-set-up-minibatches-on-one-specific-dimension-when-variables-have-multiple-and-different-dimensions/11509

```python
import pymc as pm
import numpy as np

A = rng.normal(size=(1000, 100))
B = rng.normal(size=(1000, 100))

mA, mB = pm.Minibatch(A, B, batch_size=10)
mA.eval().shape
# AssertionError: All variables shape[0] in Minibatch should be equal, check your Minibatch(data1, data2, ...) code
```

This test
https://github.com/pymc-devs/pymc/blob/ebd42a5bce64af4abc961f64f815a0f17929e16e/pymc/tests/test_data.py#L737-L742

Actually raises the same error when
```python
d1, d2 = pm.Minibatch(self.data, self.data, batch_size=20) 
```

Related to #6554

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -170,13 +170,11 @@ def assert_all_scalars_equal(scalar, *scalars):
     else:
         return Assert(
             "All variables shape[0] in Minibatch should be equal, check your Minibatch(data1, data2, ...) code"
-        )(scalar, pt.all([scalar == s for s in scalars]))
+        )(scalar, pt.all([pt.eq(scalar, s) for s in scalars]))
 
 
 def Minibatch(variable: TensorVariable, *variables: TensorVariable, batch_size: int):
-    """
-    Get random slices from variables from the leading dimension.
-
+    """Get random slices from variables from the leading dimension.
 
     Parameters
     ----------
@@ -190,6 +188,9 @@ def Minibatch(variable: TensorVariable, *variables: TensorVariable, batch_size: 
     >>> data2 = np.random.randn(100, 20)
     >>> mdata1, mdata2 = Minibatch(data1, data2, batch_size=10)
     """
+
+    if not isinstance(batch_size, int):
+        raise TypeError("batch_size must be an integer")
 
     tensor, *tensors = tuple(map(pt.as_tensor, (variable, *variables)))
     upper = assert_all_scalars_equal(*[t.shape[0] for t in (tensor, *tensors)])

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -36,6 +36,7 @@ from pytensor.tensor.var import TensorConstant, TensorVariable
 
 import pymc as pm
 
+from pymc.logprob.abstract import _get_measurable_outputs
 from pymc.pytensorf import convert_observed_data
 
 __all__ = [
@@ -132,6 +133,11 @@ class MinibatchIndexRV(IntegersRV):
         if rng is None:
             rng = pytensor.shared(np.random.default_rng())
         return super().make_node(rng, *args, **kwargs)
+
+
+@_get_measurable_outputs.register(MinibatchIndexRV)
+def minibatch_index_rv_measuarable_outputs(op, node):
+    return []
 
 
 minibatch_index = MinibatchIndexRV()

--- a/pymc/logprob/basic.py
+++ b/pymc/logprob/basic.py
@@ -44,7 +44,13 @@ import pytensor
 import pytensor.tensor as pt
 
 from pytensor import config
-from pytensor.graph.basic import Variable, graph_inputs, io_toposort
+from pytensor.graph.basic import (
+    Constant,
+    Variable,
+    ancestors,
+    graph_inputs,
+    io_toposort,
+)
 from pytensor.graph.op import compute_test_value
 from pytensor.graph.rewriting.basic import GraphRewriter, NodeRewriter
 from pytensor.tensor.random.op import RandomVariable
@@ -231,10 +237,16 @@ def factorized_joint_logprob(
     # node.
     replacements = updated_rv_values.copy()
 
-    # To avoid cloning the value variables, we map them to themselves in the
-    # `replacements` `dict` (i.e. entries already existing in `replacements`
-    # aren't cloned)
-    replacements.update({v: v for v in rv_values.values()})
+    # To avoid cloning the value variables (or ancestors of value variables),
+    # we map them to themselves in the `replacements` `dict`
+    # (i.e. entries already existing in `replacements` aren't cloned)
+    replacements.update(
+        {
+            v: v
+            for v in ancestors(rv_values.values())
+            if (not isinstance(v, Constant) and v not in replacements)
+        }
+    )
 
     # Walk the graph from its inputs to its outputs and construct the
     # log-probability

--- a/pymc/logprob/rewriting.py
+++ b/pymc/logprob/rewriting.py
@@ -39,7 +39,7 @@ from typing import Dict, Optional, Tuple
 import pytensor.tensor as pt
 
 from pytensor.compile.mode import optdb
-from pytensor.graph.basic import Variable
+from pytensor.graph.basic import Constant, Variable, ancestors
 from pytensor.graph.features import Feature
 from pytensor.graph.fg import FunctionGraph
 from pytensor.graph.rewriting.basic import GraphRewriter, node_rewriter
@@ -316,8 +316,8 @@ def construct_ir_fgraph(
     # the old nodes to the new ones; otherwise, we won't be able to use
     # `rv_values`.
     # We start the `dict` with mappings from the value variables to themselves,
-    # to prevent them from being cloned.
-    memo = {v: v for v in rv_values.values()}
+    # to prevent them from being cloned. This also includes ancestors
+    memo = {v: v for v in ancestors(rv_values.values()) if not isinstance(v, Constant)}
 
     # We add `ShapeFeature` because it will get rid of references to the old
     # `RandomVariable`s that have been lifted; otherwise, it will be difficult

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -14,6 +14,7 @@
 
 import io
 import itertools as it
+import re
 
 import cloudpickle
 import numpy as np
@@ -614,3 +615,23 @@ class TestMinibatch:
         ):
             d1, d2 = pm.Minibatch(self.data, self.data[::2], batch_size=20)
             d1.eval()
+
+    def test_multiple_vars(self):
+        A = np.arange(1000)
+        B = np.arange(1000)
+        mA, mB = pm.Minibatch(A, B, batch_size=10)
+
+        [draw_mA, draw_mB] = pm.draw([mA, mB])
+        assert draw_mA.shape == (10,)
+        np.testing.assert_allclose(draw_mA, draw_mB)
+
+        # Check invalid dims
+        A = np.arange(1000)
+        C = np.arange(999)
+        mA, mC = pm.Minibatch(A, C, batch_size=10)
+
+        with pytest.raises(
+            AssertionError,
+            match=re.escape("All variables shape[0] in Minibatch should be equal"),
+        ):
+            pm.draw([mA, mC])

--- a/tests/variational/test_minibatch_rv.py
+++ b/tests/variational/test_minibatch_rv.py
@@ -20,6 +20,7 @@ from scipy import stats as st
 import pymc as pm
 
 from pymc import Normal, draw
+from pymc.data import minibatch_index
 from pymc.testing import select_by_precision
 from pymc.variational.minibatch_rv import create_minibatch_rv
 from tests.test_data import gen1, gen2
@@ -155,3 +156,36 @@ class TestMinibatchRandomVariable:
         mx = create_minibatch_rv(x, total_size=(10,))
         assert mx is not x
         np.testing.assert_array_equal(draw(mx, random_seed=1), draw(x, random_seed=1))
+
+    @pytest.mark.filterwarnings("error")
+    def test_minibatch_parameter_and_value(self):
+        rng = np.random.default_rng(161)
+        total_size = 1000
+
+        with pm.Model(check_bounds=False) as m:
+            AD = pm.MutableData("AD", np.arange(total_size, dtype="float64"))
+            TD = pm.MutableData("TD", np.arange(total_size, dtype="float64"))
+
+            minibatch_idx = minibatch_index(0, 10, size=(9,))
+            AD_mt = AD[minibatch_idx]
+            TD_mt = TD[minibatch_idx]
+
+            pm.Normal(
+                "AD_predicted",
+                mu=TD_mt,
+                observed=AD_mt,
+                total_size=1000,
+            )
+
+        logp_fn = m.compile_logp()
+
+        ip = m.initial_point()
+        np.testing.assert_allclose(logp_fn(ip), st.norm.logpdf(0) * 1000)
+
+        with m:
+            pm.set_data({"AD": np.arange(total_size) + 1})
+        np.testing.assert_allclose(logp_fn(ip), st.norm.logpdf(1) * 1000)
+
+        with m:
+            pm.set_data({"AD": rng.normal(size=1000)})
+        assert logp_fn(ip) != logp_fn(ip)

